python/workflows.py

```diff
@@ -1,53 +1,88 @@
 from datetime import timedelta
 from temporalio import workflow
-from temporalio.exceptions import ActivityError, CancelledError
+from temporalio.common import RetryPolicy
 import asyncio
 import typing
 
 with workflow.unsafe.imports_passed_through():
-    from activities import get_bowl, put_bowl_away, add_cereal, put_cereal_back_in_box, add_milk
+    from activities import (
+        get_bowl,
+        put_bowl_away,
+        add_cereal,
+        put_cereal_back_in_box,
+        add_milk,
+    )
+
+# Trying only once to illustrate compensations easily when activities fail.
+common_retry_policy = RetryPolicy(maximum_attempts=1)
+time_delta = timedelta(seconds=5)
 
 
 class Compensations:
-   def __init__(self, parallel_compensations=False):
-       self.parallel_compensations = parallel_compensations
-       self.compensations = []
-
-
-   def add(self, function: typing.Awaitable):
-       self.compensations.append(function)
-
-
-   def __iadd__(self, function: typing.Awaitable):
-       self.add(function)
-       return self
-  
-   async def compensate(self):
-       if self.parallel_compensations:
-           try:
-               await asyncio.gather(self.compensations, return_exceptions=True)
-           except Exception as e:
-               print('failed to compensate: %s' % e)
-       else:
+    def __init__(self, parallel_compensations=False):
+        self.parallel_compensations = parallel_compensations
+        self.compensations = []
+
+    def add(self, function: typing.Callable[..., typing.Awaitable[None]]):
+        self.compensations.append(function)
+
+    def __iadd__(self, function: typing.Callable[..., typing.Awaitable[None]]):
+        self.add(function)
+        return self
+
+    async def compensate(self):
+        if self.parallel_compensations:
+
+            def compensation_lambda(f):
+                return workflow.execute_activity(
+                    f,
+                    start_to_close_timeout=time_delta,
+                    retry_policy=common_retry_policy,
+                )
+
+            all_compensations = [compensation_lambda(c) for c in self.compensations]
+            results = await asyncio.gather(*all_compensations, return_exceptions=True)
+            for result in results:
+                if isinstance(result, Exception):
+                    print("failed to compensate: %s" % e)
```

Comment:
Like Go, consider using `workflow.logger` here and elsewhere. Regular loggers are not good for using from workflows (they duplicate their logs on replay)

Reply:
ah, thank you.